File: 51-100/59.go
package main
import (
	"os"
	"fmt"
	"strconv"
	// "bytes"
	// "reflect"
)
func Use(vals ...interface{}) {
  for _, val := range vals {
    _ = val
  }
}

func readFile(fname string) (int, []byte) {
	file, _ := os.Open(fname)
	fs, _ := file.Stat()
	data := make([]byte, fs.Size())
	count, _ := file.Read(data)
  i := 0
  print(data)
  print("\nprinting data start\n")
	for i=0; i<3; i++ {
    print(data[i])
  }
  print("\nprinting data end")
	return count, data
}
func splitStr(data []byte, delimiter byte) []int {
  two_dim := make([]int, len(data)/2)
  x, start := 0, 0
	for i := range data {
		if delimiter == data[i] {
      two_dim[x], _ = strconv.Atoi(string(data[start:i]))
			start = i+1
			x++
		}
	}
  two_dim[x], _ = strconv.Atoi(string(data[start:start+2]))
  return two_dim[:x+1]
}
func string2bin(s string) string {
  var binString string
  for _, c := range s {
    binString = fmt.Sprintf("%s%b", binString, c)
  }
  return binString
}
func bin2string(b []byte) string {
  var str string
  str = fmt.Sprintf("%x", b)
  // str = rune(b)
  return str
}
func xor(b1 byte, b2 byte) byte {
  if b1 != b2 { return 1 } 
  return 0
}
func decrypt(data []int, key string){
  bin_key := string2bin(key)
  print("\n"+key)
  print("\n"+bin_key)
  fmt.Printf("\n")
  var l int = 0
  // for o := range data {
  o := 0
  for o=0; o<=1; o++ {
    // bin_d1 := strconv.FormatInt(int64(data[o]), 2)
    bin_d1 := string2bin(string(data[o]))
    dLen := len(bin_d1)
    rbuffer := make([]byte, dLen)
    for p := range bin_d1 {
      bin_d2 := bin_d1[p]
      result := xor(bin_d2, bin_key[l])
      rbuffer[p] = result
      // fmt.Printf("%b,%b: %b\t", bin_d2, bin_key[l], result)
      l++
      if l > len(bin_key)-1 { l = 0 }
    }
    // sbuffer, _ := strconv.Atoi(string(rbuffer[:]))
    sbuffer := bin2string(rbuffer)
    fmt.Printf("data[o]: %d,\tbin_d1: %s,\trbuffer: %b,\tsbuffer: %s\n",
    data[o], bin_d1, rbuffer, sbuffer)
    Use(sbuffer)
  }
}
func main() {
	_ , data := readFile("/Users/dude/fun/project_euler/data/p059_cipher.txt")
  splitted := splitStr(data, []byte(",")[0])
  alph := "abcdefghijklmnopqrstuvwxyz"
  Use(splitted)
  decrypt(splitted, "abc")
  for u := range alph {
    for v := range alph {
      for w := range alph {
        t := string(alph[u])+string(alph[v])+string(alph[w])
        Use(t)
      }
    }
  }
}

File: 51-100/59.py
d2bin = lambda data: [bin(d)[2:] for d in data]
bin2d = lambda binary: [int(b, 2) for b in binary]
xor = lambda a,b: int(a != b)
toascii = lambda num: [ord(e) for e in num]
unascii = lambda asc: [chr(e) for e in asc]

with open('/Users/dude/fun/project_euler/data/p059_cipher.txt', 'r') as f:
  sdata = f.read().split('\n')[0].split(',')
  data = [int(d) for d in sdata]

def xor_strings(s, t) -> bytes:
    """xor two same length strings together."""
    if isinstance(s, str):
        # Text strings contain single characters
        return b"".join(chr(ord(a) ^ ord(b)) for a, b in zip(s, t))
    else:
        # Bytes objects contain integer values in the range 0-255
        return bytes([a ^ b for a, b in zip(s, t)])

def xor(a, b):
  da = bin(a)[2:];db = bin(b)[2:]
  out = str(); i = 0
  big = da;  smol = db
  if len(da) < len(db):
    big = db;smol = da
  for e in big:
    out += str(int(e ==  smol[i]))
    i += 1
    if i > len(smol)-1: i = 0
  ret = int(out, 2)
  # print(f'{da} < da\n{db} < db\n{out} < out \n ret {ret}')
  return ret

def decrypt(key):
  key = toascii(key)
  i = 0
  xored = list()
  for d in data:
    xored.append(xor(d, key[i]))
    i += 1
    if i > len(key)-1: i=0
  res = ''.join(unascii(xored))
  common_words = [" the ", " and ", " of ", " for ", " is "]
  matched = [1 for ew in common_words if ew in res]
  # resplit = res.split(' ')

  if len(matched) >= 1:
    print("---------------------------------------")
    print(res)
    print("---------------------------------------")

if __name__ == "__main__":
  print(xor(42, 65))
  print(xor(107, 42))
  key = int(''.join([str(v) for v in toascii('abc')]))
  print("key: ", key)
  # print(xor_strings(data, key))

  alph = 'abcdefghijklmnopqrstuvwxyz'
  for a in alph:
    for b in alph:
      for c in alph:
        decrypt(a+b+c)

